kolibri/plugins/coach/assets/src/views/exams-page/change-exam-visibility-modal.vue

```diff
@@ -73,21 +74,69 @@
     },
     data() {
       return {
-        classSelected: false,
-        groupsSelected: [],
+        classIsSelected: this.classInitiallySelected(),
+        groupsSelected: this.initiallySelectedGroups(),
       };
     },
     computed: {
       groupOptions() {
-        return this.classGroups.map(group => ({ label: group.name, value: group.id }));
+        return this.classGroups.map(group => ({ label: group.name, id: group.id }));
       },
     },
     methods: {
+      classInitiallySelected() {
+        if (this.examVisibility.class) {
+          return true;
+        }
+        return false;
+      },
+      initiallySelectedGroups() {
+        return this.examVisibility.groups.map(
+            group => ({ label: group.collection.name, id: group.collection.id }));
+      },
+      handleSelectChange() {
+        this.classIsSelected = !this.groupsSelected.length;
+      },
       deselectGroups() {
         this.groupsSelected = [];
       },
-      deselectClass() {
-        this.classSelected = false;
+      updateVisibility() {
+        if (this.classIsSelected) {
+          if (this.classIsSelected === this.classInitiallySelected()) {
+            this.close();
+            return;
+          }
+          const classCollection = [{
+            id: this.classId,
+            name: this.className,
+            kind: CollectionKinds.CLASSROOM
+          }];
+          const groupAssignments = this.examVisibility.groups.map(
+            assignment => assignment.assignmentId);
+          this.updateExamAssignments(this.examId, classCollection, groupAssignments);
+        } else if (this.groupsSelected.length) {
+          const unassignGroups = this.initiallySelectedGroups().filter(
+              initialGroup => !this.groupsSelected.find(newGroup => newGroup.id === initialGroup.id));
+          const assignGroups = this.groupsSelected.filter(
+              newGroup => !this.initiallySelectedGroups().find(
+                  initialGroup => initialGroup.id === newGroup.id));
+
+          if (!unassignGroups.length && !assignGroups.length) {
+            this.close();
+            return;
+          }
+          const assignGroupCollections = assignGroups.map(group => ({
+            id: group.id,
+            name: group.label,
+            kind: CollectionKinds.LEARNERGROUP,
+          }));
+          let unassignments = unassignGroups.map(unassignGroup => this.examVisibility.groups.find(
+            group => group.collection.id === unassignGroup.id).assignmentId);
+          if (this.examVisibility.class) {
+            unassignments = unassignments.concat(this.examVisibility.class.assignmentId);
+          }
+          this.updateExamAssignments(this.examId, assignGroupCollections, unassignments);
```

Comment:
My suggestion was to *return* the promise from the action function, and then use it here to initiate the modal close.

```
this.updateExamAssignments(this.examId, assignGroupCollections, unassignments).then(this.close);
```